File: ports/lua/src/unicode_lua/security/calculus.lua
-- Shared verdict vocabulary for the Security Conformance Layer.
--
-- Per-family modules refine this shared vocabulary into family-specific verdict
-- structures.  A `Family` value is its stable slug string (unique per family),
-- so `family_slug` is the identity on these constants.

local M = {}

-- The detector families, in aggregator-walk / priority order.  Each value is
-- the stable reason-code slug.
M.Family = {
  MalformedUtf8 = "malformed-utf8",
  MalformedUtf16 = "malformed-utf16",
  MalformedUtf32 = "malformed-utf32",
  TagBlockPayload = "tag-block-payload",
  VariationSelectorPayload = "variation-selector-payload",
  ZeroWidthPayload = "zero-width-payload",
  SurrogateReassembly = "surrogate-reassembly",
  BidiControlBalance = "bidi-control-balance",
  NoncharacterControl = "noncharacter-control",
  HomoglyphConfusable = "homoglyph-confusable",
  MixedScriptAdmissibility = "mixed-script-admissibility",
  EmojiZwjIntegrity = "emoji-zwj-integrity",
  SkinToneVariationForgery = "skin-tone-variation-forgery",
  SourceDisplayDivergence = "source-display-divergence",
  FilenameDisguise = "filename-disguise",
  RtlInjection = "rtl-injection",
  RendererDivergence = "renderer-divergence",
  NormalizationBomb = "normalization-bomb",
  StreamSafeViolation = "stream-safe-violation",
  LocaleCaseInversion = "locale-case-inversion",
  CaseExpansionMismatch = "case-expansion-mismatch",
  WidthClassConfusion = "width-class-confusion",
  NfcIdempotenceWitness = "nfc-idempotence-witness",
  IdentifierFormDrift = "identifier-form-drift",
  CovertDisplayCompound = "covert-display-compound",
  ConfusableBidiCompound = "confusable-bidi-compound",
  AdmissibilityFormDrift = "admissibility-form-drift",
  Bip39Canonical = "bip39-canonical",
  HashInputStability = "hash-input-stability",
  AiWatermarkDetectability = "ai-watermark-detectability",
}

-- Ordered severity vocabulary: Informational < Low < Moderate < High < Critical.
M.Severity = {
  Informational = 0,
  Low = 1,
  Moderate = 2,
  High = 3,
  Critical = 4,
}

-- Five-tier adversary capability hierarchy (A0 passive .. A4 model-adaptive).
M.AdversaryTier = {
  A0 = 0,
  A1 = 1,
  A2 = 2,
  A3 = 3,
  A4 = 4,
}

-- The verdict kind, independent of any family-specific sub-threat payload.
M.ClassificationKind = {
  Clear = "Clear",
  Hazard = "Hazard",
  Compound = "Compound",
  Informational = "Informational",
}

-- The default severity associated with each classification kind.
function M.default_severity(kind)
  local K = M.ClassificationKind
  local S = M.Severity
  if kind == K.Clear then
    return S.Informational
  elseif kind == K.Hazard then
    return S.Moderate
  elseif kind == K.Compound then
    return S.High
  elseif kind == K.Informational then
    return S.Informational
  else
    error("default_severity: unknown ClassificationKind " .. tostring(kind))
  end
end

return M

File: ports/elixir/lib/unicode_security/calculus.ex
defmodule UnicodeSecurity.Calculus do
  @moduledoc """
  Shared verdict vocabulary for the Security Conformance Layer.

  Per-family modules import this vocabulary and refine it into family-specific
  verdict structures. Enum vocabularies are atoms; the `Family` order is the
  order in which the aggregator walks the detectors and the priority order
  callers can rely on when composing verdicts.
  """

  @typedoc "Detector family. Order matches the aggregator walk order."
  @type family ::
          :malformed_utf8
          | :malformed_utf16
          | :malformed_utf32
          | :tag_block_payload
          | :variation_selector_payload
          | :zero_width_payload
          | :surrogate_reassembly
          | :bidi_control_balance
          | :noncharacter_control
          | :homoglyph_confusable
          | :mixed_script_admissibility
          | :emoji_zwj_integrity
          | :skin_tone_variation_forgery
          | :source_display_divergence
          | :filename_disguise
          | :rtl_injection
          | :renderer_divergence
          | :normalization_bomb
          | :stream_safe_violation
          | :locale_case_inversion
          | :case_expansion_mismatch
          | :width_class_confusion
          | :nfc_idempotence_witness
          | :identifier_form_drift
          | :covert_display_compound
          | :confusable_bidi_compound
          | :admissibility_form_drift
          | :bip39_canonical
          | :hash_input_stability
          | :ai_watermark_detectability

  @families [
    :malformed_utf8,
    :malformed_utf16,
    :malformed_utf32,
    :tag_block_payload,
    :variation_selector_payload,
    :zero_width_payload,
    :surrogate_reassembly,
    :bidi_control_balance,
    :noncharacter_control,
    :homoglyph_confusable,
    :mixed_script_admissibility,
    :emoji_zwj_integrity,
    :skin_tone_variation_forgery,
    :source_display_divergence,
    :filename_disguise,
    :rtl_injection,
    :renderer_divergence,
    :normalization_bomb,
    :stream_safe_violation,
    :locale_case_inversion,
    :case_expansion_mismatch,
    :width_class_confusion,
    :nfc_idempotence_witness,
    :identifier_form_drift,
    :covert_display_compound,
    :confusable_bidi_compound,
    :admissibility_form_drift,
    :bip39_canonical,
    :hash_input_stability,
    :ai_watermark_detectability
  ]

  @doc "The detector families in aggregator/priority order."
  @spec families() :: [family()]
  def families, do: @families

  @typedoc "Ordered severity vocabulary: informational < low < moderate < high < critical."
  @type severity :: :informational | :low | :moderate | :high | :critical

  @doc "Numeric rank of a severity (0..4)."
  @spec severity_value(severity()) :: 0..4
  def severity_value(:informational), do: 0
  def severity_value(:low), do: 1
  def severity_value(:moderate), do: 2
  def severity_value(:high), do: 3
  def severity_value(:critical), do: 4

  @typedoc "Five-tier adversary capability hierarchy A0..A4."
  @type adversary_tier :: :a0 | :a1 | :a2 | :a3 | :a4

  @doc "Numeric rank of an adversary tier (0..4)."
  @spec adversary_tier_value(adversary_tier()) :: 0..4
  def adversary_tier_value(:a0), do: 0
  def adversary_tier_value(:a1), do: 1
  def adversary_tier_value(:a2), do: 2
  def adversary_tier_value(:a3), do: 3
  def adversary_tier_value(:a4), do: 4

  @typedoc "The verdict kind, independent of any family-specific sub-threat payload."
  @type classification_kind :: :clear | :hazard | :compound | :informational

  @doc "Default severity associated with each classification kind."
  @spec default_severity(classification_kind()) :: severity()
  def default_severity(:clear), do: :informational
  def default_severity(:hazard), do: :moderate
  def default_severity(:compound), do: :high
  def default_severity(:informational), do: :informational
end
